experimental/sql/dialects/rel_ssa.py

```diff
@@ -0,0 +1,320 @@
+
+  Example:
+
+  '''
+  %0 : rel_ssa.int32 = rel_ssa.column() ["col_name" = "id"]
+  '''
+  """
+  name = "rel_ssa.column"
+
+  col_name = AttributeDef(StringAttr)
+  result = ResultDef(DataType)
+
+
+@irdl_op_definition
+class Compare(Expression):
+  """
+  Returns `left` 'comparator' `right`.
+
+  Example:
+
+  '''
+  ...
+  %0 : !rel_ssa.bool = rel_ssa.compare(%a : !rel_ssa.int32, %b : !rel_ssa.int32) ["comparator" = "="]
+  '''
+  """
+  name = "rel_ssa.compare"
+
+  left = OperandDef(DataType)
+  right = OperandDef(DataType)
+  comparator = AttributeDef(StringAttr)
+
+  result = ResultDef(Boolean)
+
+  @staticmethod
+  @builder
+  def get(left: Operation, right: Operation,
+          comparator: StringAttr) -> 'Compare':
+    return Compare.build(operands=[left, right],
+                         attributes={"comparator": comparator},
+                         result_types=[Boolean])
+
+
+@irdl_op_definition
+class Yield(Expression):
+  """
+  Bridges the gap from expressions back to operators by yielding the result of
+  an expression to the encompassing operator.
+
+  Example:
+
+  '''
+  rel_ssa.yield(%0 : !rel_ssa.bool)
+  '''
+  """
+  name = "rel_ssa.yield"
+
+  ops = VarOperandDef(AnyAttr())
+
+  @staticmethod
+  @builder
+  def get(ops: list[Operation]) -> 'Yield':
+    return Yield.build(operands=[ops])
+
+
+@irdl_op_definition
+class Literal(Expression):
+  """
+  Creates a literal with value `value`.
+
+  Example:
+
+  '''
+  %0 : !rel_ssa.int32 = rel_ssa.literal() ["value" = 5 : i32]
+  '''
+  """
+  name = "rel_ssa.literal"
+
+  # TODO: change IntegerAttr s.t. it can have type !rel.int32
+  value = AttributeDef(AnyAttr())
+  result = ResultDef(DataType)
+
+  @staticmethod
+  @builder
+  def get(val: Attribute, res: DataType) -> 'Literal':
+    return Literal.build(attributes={"value": val}, result_types=[res])
+
+
+#===------------------------------------------------------------------------===#
+# Operators
+#===------------------------------------------------------------------------===#
+
+
+class Operator(Operation):
+  """
+  Interface class for all Operators, i.e., operations that work on bags.
+  """
+  ...
+
+
+@irdl_op_definition
+class PandasTable(Operator):
+  """
+  Defines a table with name `table_name`.
+
+  Example:
+
+  '''
+  %0 : rel_ssa.bag<[!rel_ssa.int32]> = rel_ssa.pandas_table() ["table_name" = "t"]
+  '''
+  """
+  name = "rel_ssa.pandas_table"
+
+  table_name = AttributeDef(StringAttr)
+  result = ResultDef(Bag)
+
+  @staticmethod
+  @builder
+  def get(name: str, result_type: Attribute) -> 'PandasTable':
+    return PandasTable.build(
+        attributes={"table_name": StringAttr.from_str(name)},
+        result_types=[result_type])
+
+
+@irdl_op_definition
+class Select(Operator):
+  """
+  Selects all tuples of `input` according to the yielded expression in `predicates`.
+
+  Example:
+
+  '''
+  rel_ssa.select(%0: rel_ssa.bag<[!rel_ssa.int32]>) {
```

Comment:
I am quite unclear about wanting this modeling as it seems you could achieve the same with:
```
func.func some_fun(%0: i32) -> i1 {
  ...
  return %3 : i1
}

rel_ssa.select "some_fun" from %0: rel_ssa.bag<i32>
```

Reply:
Yes, I think you can achieve the same using this modeling. I must admit, I don't feel like I understand all the details well enough to really decide what makes most sense. To me, the current version is nice, since the operation and the function that the operation applies are very close to each other. That feels to me like it makes the IR more readable than having some functions at first and then all the actual `select` operations.

I think in a lowered setting, say the `iterators`-dialect, we might want to have this function version. But for this, we can just move the body to another function during the lowering, I feel.